src/mono/wasm/runtime/crypto-worker.ts

```diff
@@ -299,33 +291,39 @@ class LibraryChannel {
     private read_response(): string {
         let response = "";
         for (; ;) {
-            const state = this.wait_for_state(state => state == this.STATE_RESP || state == this.STATE_RESP_P, "read_response");
-            this.acquire_lock();
-
-            try {
+            this.wait_for_state_change_to(state => state == this.STATE_RESP || state == this.STATE_RESP_P, "read_response");
+            let state;
+            const done = this._using_lock_for_nonblocking_cb(() => {
                 const size_to_read = Atomics.load(this.comm, this.MSG_SIZE_IDX);
 
                 // Append the latest part of the message.
                 response += this.read_from_msg(0, size_to_read);
 
                 // The response is complete.
+                state = Atomics.load(this.comm, this.STATE_IDX);
```

Comment:
Why do we need this extra read of the state? Once the state is `STATE_RESP || STATE_RESP_P`, the only thing that can (should?) change the state is the main thread (i.e. this code).